Sources/NIOCore/AsyncSequences/NIOBackPressuredAsyncSequence.swift

```diff
@@ -0,0 +1,1004 @@
+        }
+
+        @inlinable
+        mutating func yield<S: Sequence>(_ sequence: S) -> YieldAction where S.Element == Element {
+            switch self.state {
+            case .initial(var backPressureStrategy, let iteratorInitialized):
+                self.state = .modifying
+
+                let buffer = CircularBuffer<Element>(sequence)
+                let shouldDemandMore = backPressureStrategy.didYield(bufferDepth: buffer.count)
+
+                self.state = .streaming(
+                    backPressureStrategy: backPressureStrategy,
+                    buffer: buffer,
+                    continuation: nil,
+                    hasOutstandingDemand: shouldDemandMore,
+                    iteratorInitialized: iteratorInitialized
+                )
+
+                if shouldDemandMore {
+                    return .returnDemandMore
+                } else {
+                    return .returnStopDemanding
+                }
+
+            case .streaming(var backPressureStrategy, var buffer, .some(let continuation), _, let iteratorInitialized):
+                // The buffer should always be empty if we hold a continuation
+                precondition(buffer.isEmpty)
+
+                self.state = .modifying
+
+                buffer.append(contentsOf: sequence)
+                let element = buffer.popFirst()
+                let shouldDemandMore = backPressureStrategy.didYield(bufferDepth: buffer.count)
+
+                if let element = element {
+                    // We have an element and can resume the continuation
+                    self.state = .streaming(
+                        backPressureStrategy: backPressureStrategy,
+                        buffer: buffer,
+                        continuation: nil, // Setting this to nil since we are resuming the continuation
+                        hasOutstandingDemand: shouldDemandMore,
+                        iteratorInitialized: iteratorInitialized
+                    )
+
+                    if shouldDemandMore {
+                        return .resumeContinuationAndReturnDemandMore(
+                            continuation: continuation,
+                            element: element
+                        )
+                    } else {
+                        return .resumeContinuationAndReturnStopDemanding(
+                            continuation: continuation,
+                            element: element
+                        )
+                    }
+                } else {
+                    self.state = .streaming(
+                        backPressureStrategy: backPressureStrategy,
+                        buffer: buffer,
+                        continuation: continuation,
+                        hasOutstandingDemand: shouldDemandMore,
+                        iteratorInitialized: iteratorInitialized
+                    )
+
+                    // This is weird somebody yielded an empty sequence but we can handle it
```

Comment:
I thought about simplifying this before already however I think regardless of how you change it around it won't get much shorter. I don't think we can return early with the current `hasOutstandingDemand` value. We need to inform the strategy even if an empty `yield` happened. Furthermore, the strategy might be doing any weird kind of thing where it returns a different demand behaviour even-though the `bufferDepth` is exactly the same.

I reordered this a bit a couple of times already and it always results in similar amounts of code.

Reply:
Why do we need to inform the strategy on empty yield?

Reply:
IMO if the producer `yielded` to the source he is expecting that the strategy he configured is getting called and has a chance to evaluate with the current `bufferDepth`, if for some weird reason somebody implemented a strategy that signals demand on every other `yield()` invocation than early exiting on empty yields would not work.
In general, I think we are not losing anything by calling the strategy on empty yields, rather, we are just avoiding bugs down the road. Moreover, the empty `yield()` case is probably not happing too often anyhow.

Reply:
I think the exact opposite is true here.

Users don't have _any_ expectations about this type, because this type doesn't exist today and is not a direct analog of anything that currently exists. We define what the semantics of the `yield` operation are as it relates to the strategy. It is deeply not obvious to me that we might call the `didYield` method when nothing was actually yielded.